src/Adapter/Product/ProductDuplicator.php

```diff
@@ -0,0 +1,535 @@
+    }
+
+    /**
+     * @param Product $newProduct
+     * @param int $oldProductId
+     *
+     * @throws CannotUpdateProductException
+     */
+    private function updateSearchIndexation(Product $newProduct, int $oldProductId): void
+    {
+        if (!in_array($newProduct->visibility, ['both', 'search']) || !$this->isSearchIndexationOn) {
+            return;
+        }
+
+        try {
+            if (!Search::indexation(false, $newProduct->id)) {
+                throw new CannotUpdateProductException(
+                    sprintf('Cannot update search indexation when duplicating product %d', $oldProductId),
+                    CannotUpdateProductException::FAILED_UPDATE_SEARCH_INDEXATION
+                );
+            }
+        } catch (PrestaShopException $e) {
+            throw new CoreException(
+                sprintf('Error occurred when duplicating product %d. Failed to update search indexation', $oldProductId),
+                0,
+                $e
+            );
+        }
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @return Product the new product
+     */
+    private function duplicateProduct(Product $product): Product
+    {
+        $this->setName($product);
+        $this->setPriceByShops($product);
+
+        unset($product->id, $product->id_product);
+        $product->indexed = false;
+        $product->active = false;
+
+        $this->productRepository->add($product)->getValue();
+
+        return $product;
+    }
+
+    /**
+     * Sets new name for duplicated product by modifying the existing name with a new pattern
+     *
+     * @param Product $product
+     */
+    private function setName(Product $product): void
+    {
+        //@todo: would it be possible to transle it or is it ok that each lang will contain this generic pattern?
+        //@todo: make the pattern modifiable (add configuration) ?
+        $namePattern = 'copy of %s';
+
+        foreach ($product->name as $langKey => $oldName) {
+            $newName = sprintf($namePattern, $oldName);
+            $product->name[$langKey] = $newName;
+            //@todo: validate name if it is too long now. In that case, cut the name or throw exception?
+        }
+    }
+
+    /**
+     * @param Product $product
+     */
+    private function setPriceByShops(Product $product): void
+    {
+        if (!empty($product->price) || !$this->multistoreContextChecker->isGroupShopContext()) {
+            return;
+        }
+
+        foreach ($this->getContextShops() as $shop) {
+            $priceByShop = $this->productRepository->getPriceByShop(
+                new ProductId((int) $product->id),
+                new ShopId((int) $shop['id_shop'])
+            );
+
+            if (!$priceByShop) {
+                continue;
+            }
+
+            $product->price = (string) $priceByShop;
+        }
+    }
+
+    /**
+     * @todo: move this somewhere? add methods in MultistoreContextCheckerInterface?
```

Comment:
Just this `ShopGroup::getShopsFromGroup(Shop::getContextShopGroupID())` ? It does not look too bad / too long, I think we can keep it here.

Reply:
Agreed besides `MultistoreContextCheckerInterface` is supposed to check a state, not get data depending on the shop context, this would be managed by another interface, or this one should be renamed